nixos/modules/config/users-groups.nix

```diff
@@ -354,17 +354,32 @@ let
     };
   };
 
-  mkSubuidEntry = user: concatStrings (
-    map (range: "${user.name}:${toString range.startUid}:${toString range.count}\n")
-      user.subUidRanges);
+  # format a list of lists into the colon-separated lines of /etc/passwd and similar
+  mkDatabase = concatMapStrings (entry: concatMapStringsSep ":" toString entry + "\n");
 
-  subuidFile = concatStrings (map mkSubuidEntry (attrValues cfg.users));
+  accountsAreDeterministic = !cfg.mutableUsers && cfg.enforceStaticIds;
```

Comment:
As I understand it, with this check it won't build static /etc/{shadow,groups} if you have `mutableUsers = true`. I think it would be better to not have any detection for this, just use the value of the option directly to determine whether it should be static or not, and have an assertion verify that `mutableUsers = false`. Because why would anybody want to turn on this option if it doesn't end up doing anything

Reply:
The assertion which `enforceStaticIds` now turns on (all `uid`s/`gid`s are non-`null`) still applies even if `mutableUsers = true`. I'm not certain whether anyone would care about the assertion in that case, but it does do something.

I'm having an easier time thinking about explaining this feature when implemented this way than if the new option directly controls the static file generation. "Here's what `mutableUsers` does, and here's a property of IDs which we could enforce for you, and if you put those together then you get this other thing" feels simpler than starting with static file generation and then explaining the preconditions on it.

But I don't feel strongly either way, especially if somebody has a better way to document this.